src/Lecture2.hs

```diff
@@ -51,8 +51,14 @@ zero, you can stop calculating product and return 0 immediately.
 >>> lazyProduct [4, 3, 7]
 84
 -}
-lazyProduct :: [Int] -> Int
-lazyProduct = error "TODO"
+lazyProduct :: (Num a, Eq a) => [a] -> a
+lazyProduct list = calculateSum 1 list
+          where
+            calculateSum :: (Num a, Eq a) => a -> [a] -> a
+            calculateSum sum l
+              | l == [] = sum
+              | head l == 0 = 0
+              | otherwise = calculateSum (sum * head l) (tail l)
```

Comment:
Instead of comparing the list with an empty list and then using unsafe functions `head` and `tail`, it's better to do pattern matching on the list.

Also, `sum` is the name of the standard function in Haskell. It works because Haskell allows name shadowing. But usually it's better to avoid conflicts with with standard names 🙂